src/state/delete-text.action.ts

```diff
@@ -0,0 +1,185 @@
+                    endSegmentIndex: 0,
+                    endOffset: 0
+                });
+            }
+            
+        } else {
+            newCursor = Object.assign({}, newCursor, {
+                startOffset: newCursor.startOffset - 1,
+                endOffset: newCursor.endOffset - 1
+            });
+        }
+
+        return Object.assign({}, state, {
+            segments: newSegments,
+            cursor: newCursor
+        });
+    }
+
+    private deleteMultiple(state: IEditorState): IEditorState {
+        const { cursor, segments } = state;
+        let newSegments: ISegment[] = segments;
+        let newCursor: ICursor = cursor;
+
+        let startSegment = findSegment(cursor.startSegmentIndex, segments);
+        if (!startSegment) {
+            return state;
+        }
+
+        let endSegment = findSegment(cursor.endSegmentIndex, segments);
+        if (!endSegment) {
+            return state;
+        }
+
+        startSegment = Object.assign<any, ISegment, Partial<ISegment>>(
+            {},
+            startSegment,
+            {
+                // to remove the single char at the cursor
+                content: remove<string>(
+                    startSegment.content,
+                    cursor.startOffset,
+                    startSegment.content.length - cursor.startOffset
+                )
+            }
+        );
+
+        endSegment = Object.assign<any, ISegment, Partial<ISegment>>(
+            {},
+            endSegment,
+            {
+                // to remove the single char at the cursor
+                content: remove<string>(
+                    endSegment.content,
+                    0,
+                    cursor.endOffset
+                )
+            }
+        );
+
+        newSegments = newSegments.map((currSegment: ISegment) => {
+            if (currSegment.index === startSegment.index) {
+                return startSegment;
+            }
+            if (currSegment.index === endSegment.index) {
+                return endSegment;
+            }
+            return currSegment;
+        });
+
+        let startIndex = startSegment.index;
+        let endIndex = endSegment.index;
+        if (startSegment.content.length === 0) {
+            startIndex = startIndex - 1;
+        }
+        if (endSegment.content.length === 0) {
+            endIndex = endIndex + 1;
+        }
+
+        if (endIndex - startIndex - 1 > 0) {
+            newSegments = remove<ISegment>(newSegments, startIndex + 1, endIndex - startIndex - 1);
+            newSegments = this.rematchIndex(newSegments);
+        }
+        
+        newCursor = Object.assign({}, newCursor, {
+            endSegmentIndex: newCursor.startSegmentIndex,
+            endOffset: newCursor.startOffset
+        });
+
+        return Object.assign({}, state, {
+            segments: newSegments,
+            cursor: newCursor
+        });
+    }
+
+    public perform(state: IEditorState): IEditorState {
```

Comment:
Let's move public function to the top so that it's easier for readers to follow the logic.